source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveActionSetRenderer.cs

```diff
@@ -13,34 +13,40 @@ public static class AdaptiveActionSetRenderer
     {
         public static FrameworkElement Render(AdaptiveActionSet actionSet, AdaptiveRenderContext context)
         {
-            var outerActionSet = new Grid();
+            Grid outerActionSet = new Grid();
 
             if (!context.Config.SupportsInteractivity)
                 return outerActionSet;
 
             outerActionSet.Style = context.GetStyle("Adaptive.Container");
 
             // Keep track of ContainerStyle.ForegroundColors before Container is rendered
-            var parentRenderArgs = context.RenderArgs;
-            var elementRenderArgs = new AdaptiveRenderArgs(parentRenderArgs);
+            AdaptiveRenderArgs parentRenderArgs = context.RenderArgs;
+            AdaptiveRenderArgs elementRenderArgs = new AdaptiveRenderArgs(parentRenderArgs);
 
-            AddActions(outerActionSet, actionSet.Actions, context);
+            RenderActions(outerActionSet, actionSet.Actions, context);
 
             return outerActionSet;
         }
 
-        public static void AddActions(Grid uiContainer, IList<AdaptiveAction> actions, AdaptiveRenderContext context)
+        public static void AddRenderedActions(Grid uiContainer, IList<AdaptiveAction> actions, AdaptiveRenderContext context)
+        {
+            RenderActions(uiContainer, actions, context);
+            AddActionsToRoot(context);
```

Comment:
you could change this function call to the last lines of RenderActions to avoid calling explicitly AddActionsToRoot every time RenderActions is called

Reply:
it was splited by design. the order in which view for show card is added is important. render, for example, should not call AddActionsToRoot after the call to RenderActions becauase we want to add the showcard view after all buttons in their actionset are added to the root view.

---
In reply to: [284909621](https://github.com/microsoft/AdaptiveCards/pull/2927#discussion_r284909621) [](ancestors = 284909621)